### PR TITLE
Introduce builddef.VersionMap struct

### PR DESCRIPTION
**Pull request by NiR-**
This new struct is used in specialized definitions whenever a map of packages/extensions/etc... has to be associated with version constraints. This struct has an `Add()` method which is used to define values without overwriting manually-defined ones. This was originally motivated by a bug in the php handler but this struct seems useful in other places.

Here's the original bug description:

> The infered extensions were overwriting manually defined version
constraints. Because of that, PHP prod stages were always resolving the
latest version of APCu, even when a specific version was pinned in the
zbuildfile.
>
> This new struct is used to properly add infered extensions without
overwriting manually defined version constraints.

Supersedes #20.

### CODE REVIEW

File: pkg/defkinds/php/definition.go [hunk 1]
```diff
@@ -595,83 +578,68 @@ func addIntegrations(stageDef *StageDefinition) error {
 //
 // This list has been obtained using:
 // docker run --rm -t php:7.2-fpm-buster php -r 'var_dump(get_loaded_extensions());'
-var preinstalledExtensions = []string{
```
Comment (nm2107) [inline pkg/defkinds/php/definition.go]: what about also update the above comment to indicate what the struct is supposed to contain ?

File: pkg/defkinds/php/definition.go [hunk 2]
```diff
@@ -595,83 +578,68 @@ func addIntegrations(stageDef *StageDefinition) error {
...
 }
 
 func inferExtensions(def *StageDefinition) {
 	// soap extension needs sockets extension to work properly
-	if _, ok := def.Extensions["soap"]; ok {
-		if _, ok := def.Extensions["sockets"]; !ok {
-			def.Extensions["sockets"] = "*"
-		}
+	if def.Extensions.Has("soap") {
+		def.Extensions.Add("sockets", "*", false)
```
Comment (nm2107) [inline pkg/defkinds/php/definition.go]: :+1:

File: pkg/builddef/builddef.go [hunk 3]
```diff
@@ -16,3 +16,78 @@ type BuildDef struct {
...
 	RawLocks() ([]byte, error)
 }
+
+// VersionMap is a list of packages/extensions/etc... associated to version
+// constraints or resolved versions. It has dedicated methods to manipulate
+// items in the map, most notably Add() which can be used to add an item
+// without overwriting any preexisting value. This is used in specialized
+// defkind handlers to add infered items without rewriting manually-defined
+// values.
+type VersionMap map[string]string
```
Comment (nm2107) [inline pkg/builddef/builddef.go]: what about writing unit tests for this new struct and its impl ?

File: pkg/defkinds/php/definition.go (hunk 1 above)
Comment (NiR-) [reply pkg/defkinds/php/definition.go]: Mh I can but using a map with empty struct/interface as value is quite a usual pattern in golang whenever a slice starts to be too big :wink:

**Comment by NiR-** [comment]
I added a commit which is introducing a new method named `Overwrite()`. It replaces the `overwrite` bool at the end of `Add()` method.

I also added proper tests for Add, Overwrite and Remove methods but not for others. There're already a lot of tests in the app (~7k loc of tests for ~5k of real code) and the others methods are called from defkind handlers so if they break, other parts of the code should too.

**Pull request merged**